presto-main/src/main/java/com/facebook/presto/sql/planner/optimizations/PredicatePushDown.java

```diff
@@ -599,38 +620,149 @@ public PlanNode visitJoin(JoinNode node, RewriteContext<RowExpression> context)
+            // Since we use PredicatePushdown to push dynamic filters themselves,
+            // instead of separate ApplyDynamicFilters rule we derive dynamic filters within PredicatePushdown itself.
+            // Even if equiJoinClauses.equals(node.getCriteria), current dynamic filters may not match equiJoinClauses
+            ImmutableList.Builder<CallExpression> clausesBuilder = ImmutableList.builder();
+            for (JoinNode.EquiJoinClause clause : equiJoinClauses) {
+                VariableReferenceExpression probeSymbol = clause.getLeft();
+                VariableReferenceExpression buildSymbol = clause.getRight();
+                clausesBuilder.add(call(
+                                    EQUAL.name(),
+                                    functionAndTypeManager.resolveOperator(EQUAL, fromTypes(probeSymbol.getType(), buildSymbol.getType())),
+                                    BOOLEAN,
+                                    probeSymbol,
+                                    buildSymbol));
+            }
+
+            for (RowExpression filter : joinFilter) {
+                if ((filter instanceof CallExpression)) {
+                    CallExpression call = (CallExpression) filter;
+                    List<RowExpression> arguments = call.getArguments();
+
+                    // TODO: support for complex inequalities, e.g. left < right + 10, NOT, LIKE
+                    if (arguments.size() == 1) {
+                        continue;
+                    }
+
+                    if (arguments.size() == 3) {
+                        // try convert BETWEEN into GREATER_THAN_OR_EQUAL and LESS_THAN_OR_EQUAL
+                        String function = call.getDisplayName();
+                        if (function.equals(BETWEEN.name()) && arguments.get(0) instanceof VariableReferenceExpression) {
+                            if (arguments.get(1) instanceof VariableReferenceExpression) {
+                                CallExpression callExpression = call(
+                                                                    GREATER_THAN_OR_EQUAL.name(),
+                                                                    functionAndTypeManager.resolveOperator(GREATER_THAN_OR_EQUAL, fromTypes(arguments.get(0).getType(), arguments.get(1).getType())),
+                                                                    BOOLEAN,
+                                                                    arguments.get(0),
+                                                                    arguments.get(1));
+                                Optional<CallExpression> comparisonExpression = getDynamicFilterComparison(node, callExpression, functionAndTypeManager);
+                                if (comparisonExpression.isPresent()) {
+                                    clausesBuilder.add(comparisonExpression.get());
+                                }
+                            }
```

Comment:
Recommendation generated by Amazon CodeGuru Reviewer. Leave feedback on this recommendation by replying to the comment or by reacting to the comment using emoji.

Similar code fragments were detected in the same file at the following lines: 690:701, 702:713.  
Refactoring can help improve code maintainability. Consider reducing duplicate code by extracting it into a separate method. You can then replace duplicated code with calls to this new method.